### DISCUSSION TITLE
[java] How to understand "AvoidFinalLocalVariable" and "LocalVariableCouldBeFinal"

### DISCUSSION THREAD

**Discussion by lonelyma1021** [pmd/pmd#3043] [Q&A] [answered]
The rule "AvoidFinalLocalVariable" says: Avoid using final local variables, turn them into fields. And the rule "LocalVariableCouldBeFinal" says: A local variable assigned only once can be declared final. It seems that they go against each other. The result of the two rules above is that all the local variables which assigned once turn into fields. I know that final variable can improve performance. But what's the difference between a local variable and a field? Why should I turn local variable to field?

**Comment by jsotuyod** [comment, marked as answer]
@lonelyma1021 these are separate rules. It's up to each user to decide if they apply to their projects or not (and if both are required!).

`LocalVariableCouldBeFinal` attempts to flag local variables as finals. This has 2 main purposes:
1. help the compiler / JIT improve performance
2. minimizing mutability

This last one is, in my opinion, the strongest reason to adopt this rule.

Reassigning parameters / local variables means that knowing the actual value they have at a single point requires looking at the whole code between declaration and that point, taking into account all plausible execution paths (which corresponds to cyclomatic complexity).

If variables are final, you only have to look for it's initialization, and be safe the value will be that.

Most often than not, if you are reassigning a variable it's either:
1. because you are working within a loop, which is probably ok
2. you just really want a new variable, and are being a little lazy


As for `AvoidFinalLocalVariable`, this used to be part of the controversial ruleset back in PMD 5.x. It considers that, if a final local variable is always set to the same value, it should probably be a field (actually, most probably a `static final` field). This is a matter of style, wether all such constants should be fields, even if not used in other methods or not.
